### PR TITLE
[VM] Added Native Structs to the file format

### PR DESCRIPTION
**Pull request by tnowacki**
## Motivation

* Native struct declarations will be needed utilized to incorporate collections in a seamless manner
* To begin this work on collections, I have added native structs to the file format
* To prevent code paths from even looking for fields on native modules, I made an enum for "body" of the struct definition, either `Native` and `Declared { field_count, fields }`
  * In the serialization/deserialization, `Native` is represented by a flag, and a value of 0 for both `field_count` and `fields`  
  * While other enums are used in the file format (signature tokens, kinds, bytecode instructions), this would be a bit different as it is abstracting over an invariant in the file format.
  * I attempted to instead have a 1-1 correspondence between the binary and Rust version of the file format, but this made the code more cumbersome for no gain. In other words, having a separate field for `flag`, `field_count`, and `fields` added a ton of weight with little gain. The additional complexity came in the form of: 
    * Custom implementation for proptest Arbitrary 
    * Private fields with a public getter that constructed the `StructFieldInformation` enum every call
    * A `new` function as the field count and field index were now private 

## Test Plan

* Updated Prop Tests
* cargo test
* Extensive testing will be difficult without IR support and execution support (in the form of a collection implementation)

### CODE REVIEW

File: language/bytecode_verifier/src/type_memory_safety.rs [hunk 1]
```diff
@@ -428,11 +428,21 @@ impl<'a> TypeAndMemorySafetyAnalysis<'a> {
...
                 let struct_definition_view =
                     StructDefinitionView::new(self.module, struct_definition);
-                for field_definition_view in struct_definition_view.fields().rev() {
-                    let field_signature_view = field_definition_view.type_signature();
-                    let arg = self.stack.pop().unwrap();
-                    if arg.signature != *field_signature_view.token().as_inner() {
-                        return Err(VMStaticViolation::PackTypeMismatchError(offset));
+                match struct_definition_view.fields() {
+                    None => {
+                        // TODO unpack on native error
```
Comment (sblackshear) [inline language/bytecode_verifier/src/type_memory_safety.rs]:
```suggestion
                        // TODO pack on native error
```

File: language/bytecode_verifier/src/struct_defs.rs [hunk 2]
```diff
@@ -97,7 +97,7 @@ impl<'a> StructDefGraphBuilder<'a> {
     ) -> impl Iterator<Item = StructDefinitionIndex> + 'a {
         let struct_def = self.module.struct_def_at(idx);
         let struct_def = StructDefinitionView::new(self.module, struct_def);
-        let fields = struct_def.fields();
+        let fields = struct_def.fields().into_iter().flatten();
```
Comment (sblackshear) [inline language/bytecode_verifier/src/struct_defs.rs]: It's a bit hard to tell this is dealing with the `Native` case. A comment or an explicit pattern match might help?

File: language/vm/vm_runtime/src/code_cache/module_cache.rs [hunk 3]
```diff
@@ -284,21 +287,29 @@ impl<'alloc> VMModuleCache<'alloc> {
...
-                    &module.type_signature_at(field.signature).0,
-                    gas_meter,
-                    fetcher
-                ));
-                if let Some(t) = ty {
-                    field_types.push(t);
-                } else {
-                    return Ok(Ok(None));
+            match &struct_def.field_information {
+                StructFieldInformation::Native => return Err(VMInvariantViolation::LinkerError),
```
Comment (sblackshear) [inline language/vm/vm_runtime/src/code_cache/module_cache.rs]: TODO for handling this in the future or returning a more descriptive error?

File: language/vm/src/file_format.rs [hunk 4]
```diff
@@ -251,20 +254,42 @@ pub struct FunctionHandle {
...
+    /// Contains either
+    /// - Information indicating the struct is native and has no accessible fields
+    /// - Information indicating the number of fields and the start `FieldDefinitionIndex`
+    pub field_information: StructFieldInformation,
 }
 
+impl StructDefinition {
+    pub fn declared_field_count(&self) -> Result<MemberCount, VMInvariantViolation> {
+        match &self.field_information {
+            StructFieldInformation::Native => Err(VMInvariantViolation::LinkerError),
```
Comment (sblackshear) [inline language/vm/src/file_format.rs]: Is this the right error to report here?

Comment (tnowacki) [reply language/vm/src/file_format.rs]: It's the same error we give on improper invocation of native procedures. 

I left it there as a placeholder mostly. 

@runtian-zhou thoughts?

Comment (tnowacki) [reply language/vm/src/file_format.rs]: I'll just give it a TODO maybe for now

File: language/vm/vm_runtime/src/code_cache/module_cache.rs (hunk 3 above)
Comment (tnowacki) [reply language/vm/vm_runtime/src/code_cache/module_cache.rs]: It's the same one we use for native procedures, but I'll add a TODO

**Pull request merged**